compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/viewinterop/SwingInteropContainer.desktop.kt

```diff
@@ -60,53 +67,63 @@ internal class SwingInteropContainer(
             return lastInteropIndex
         }
 
-    override fun placeInteropView(interopView: InteropViewHolder) {
-        val component = interopView.group
+    override fun contains(holder: InteropViewHolder): Boolean =
+        interopComponents.contains(holder.group)
+
+    override fun place(holder: InteropViewHolder) {
+        val group = holder.group
+
+        if (interopComponents.isEmpty()) {
+            snapshotObserver.start()
+        }
 
         // Add this component to [interopComponents] to track count and clip rects
-        val alreadyAdded = component in interopComponents
+        val alreadyAdded = group in interopComponents
         if (!alreadyAdded) {
-            interopComponents[component] = interopView
+            interopComponents[group] = holder
         }
 
         // Iterate through a Compose layout tree in draw order and count interop view below this one
-        val countBelow = countInteropComponentsBelow(interopView)
+        val countBelow = countInteropComponentsBelow(holder)
 
         // AWT/Swing uses the **REVERSE ORDER** for drawing and events
         val awtIndex = lastInteropIndex - countBelow
 
         // Update AWT/Swing hierarchy
         if (alreadyAdded) {
-            root.setComponentZOrder(component, awtIndex)
+            holder.changeInteropViewIndex(root, awtIndex)
         } else {
-            root.add(component, awtIndex)
+            holder.insertInteropView(root, awtIndex)
         }
 
-        // Sometimes Swing displays the rest of interop views in incorrect order after adding,
-        // so we need to force re-validate it.
         root.validate()
         root.repaint()
     }
 
-    override fun unplaceInteropView(interopView: InteropViewHolder) {
-        val component = interopView.group
-        root.remove(component)
-        interopComponents.remove(component)
+    override fun unplace(holder: InteropViewHolder) {
+        holder.removeInteropView(root)
+
+        interopComponents.remove(holder.group)
+
+        if (interopComponents.isEmpty()) {
+            snapshotObserver.stop()
+        }
 
-        // Sometimes Swing displays the rest of interop views in incorrect order after removing,
-        // so we need to force re-validate it.
         root.validate()
         root.repaint()
     }
 
-    override fun changeInteropViewLayout(action: () -> Unit) {
+    override fun update(action: () -> Unit) = SwingUtilities.invokeLater {
```

Comment:
Right, if there are no strong need for this, better to avoid using the global dispatcher. It:
- has unpredictable data flow
- can cause concurrency issues
- makes "dispose" to not dispose everything (there is still scheduled, "leaked" operation after "dispose").

To avoid such issues, we should
- wrap it to `InteropViewUpdater` (solves "dispose" and double update issue)
- make it the only owner of the updates/state changes (solves the concurrency issue)
- (optional) replace `SwingUtilities.invokeLater` by a custom "interop dispatcher" that is called manually inside `ComposeScene.render` (makes data flow predictable)

Reply:
Just to be on the same page:
This functions has no different semantics from what `InteropViewUpdater` did [here](https://github.com/JetBrains/compose-multiplatform-core/blob/a8ea6905428034f620b6bed41847be829e37aa1d/compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/awt/SwingPanel.desktop.kt#L299). So it's not a change.

Do we track it as an issue already?

Reply:
> So it's not a change.

It looks like it changed. There is no protection for double scheduling and disposing, which was inside `InteropViewUpdater`.

Reply:
There is disposing in:
`snapshotObserver.clear(this)` 
There is no double scheduling, because from my understanding it can scheduled to be executed only once per recomposition.

Reply:
Is there a rationale for `SwingUtilities.invokeLater` in an original implementation in a first place?
All other changes in Swing are synchronous.

Reply:
> There is disposing in: snapshotObserver.clear(this)

It isn't related to this. We need to dispose a task that we scheduled to Swing.

>  it can scheduled to be executed only once per recomposition.

- it looks like implementation details that are not obvious for the reader of this particular code
- recomposition is independent of Swing dispatcher, and can happen twice without freeing the dispatcher. A real world example - recomposition in LazyColumn happens synchronously in a scroll event.

Reply:
> Is there a rationale for SwingUtilities.invokeLater in an original implementation in a first place?

It was originally added [here](https://android-review.googlesource.com/c/platform/frameworks/support/+/1559240). Scheduling was necessary there because we used `SnapshotStateObserver` that can happen multiple times.

As we rewrote it to `Applier`, most probably we can remove scheduling.

Reply:
> Scheduling was necessary there

SwingUtilities itself wasn't necessary, we could reuse effect dispatcher that runs on UI thread (see `UpdateEffect`)

Reply:
> it looks like implementation details that are not obvious for the reader of this particular code
Absolutely not. Neither is previous.

> recomposition is independent of Swing dispatcher, and can happen twice without freeing the dispatcher. A real world example - recomposition in LazyColumn happens synchronously in a scroll event.
Yes, but updates are only dispatched if the state read by it has changed.

> it looks like it is also called in isAttachedToWindow that can be changed multiple times
Only once when view is initially added. Or update lambda is changed by the user.
Which makes sense?

> As we rewrote it to Applier, most probably we can remove scheduling.
That's my idea too. So I guess we can just remove scheduling altogether and things should be correct?

Reply:
> remove scheduling altogether and things should be correct?

Yes

Reply:
Fixed

Reply:
Sorry, I looked at the code more, and we still call it on `snapshotObserver.observeReads`. So, we need scheduling. let'me look at the code more, to understand why we use `Applier` and `snapshotObserver` at the same time

Reply:
> Applier and snapshotObserver at the same time

If Applier don't perform state observing by itself, the we indeed need it, and we need scheduling. If so, let's discuss scheduling tomorrow in the evening, I feel we have different perspectives and don't understand each other.

Reply:
> and we need scheduling

This is not only about a correct code, we can have a potential bug (because of non-cancelling) or UI glitches (because of double scheduling)

Reply:
> If Applier don't perform state observing by itsel

It it performs state observing, let's remove `snapshotObserver.observeReads`, there won't be questions/disqussions then

Reply:
> non-cancelling

No updates are called when the swing interop component is detached now.

> UI glitches

Updates will be performed only if state they read change. This is an entry-point for user-provided `update`, can you help me with working out a scenario where multiple updates can happen in-between internal recompositions?

Reply:
> No updates are called when the swing interop component is detached now.

update -> schedule -> dispose -> perform scheduled task

Reply:
> perform scheduled task

This task also can have some code that throws an exception when the native view is detached

Reply:
But scheduling was removed. It's now synchronous on Swing. So it won't be called after component is detached, due to `snapshotObserver.clear(this)`

Reply:
> me with working out a scenario where multiple updates can happen in-between internal recompositions?

It no longer about recompositions, because `snapshotObserver` is involved (initially I though that it is only called inside `Applier`). `snapshotObserver` isn't bound to recompositions or to `ComposeScene` in any way, it can be called anytime, in any thread by contract. For example, if app runs own thread and performs a state change, `onChangedExecutor` is called immediately in this thread.
![image](https://github.com/user-attachments/assets/c7c9160e-62b7-4274-b2c5-2d408a59c90b)

Reply:
> is called immediately in this thread.

So, we also have a threading issue, we need to perform Interop updates on UI thread, which is `SwingUtlities.invokeLater` or `LaunchedEffect`

Reply:
> which is SwingUtlities.invokeLater or LaunchedEffect

Or a custom dispatcher, which we call in `ComposeScene.render` to sync updates with Compose.

Or just move updates into `ComposeNode(update=`, if it can observe states. Could you check?

Reply:
> But scheduling (using invokeLater) was removed

I am talking about when we have scheduling. We need it if we use `snapshotObserver`

Reply:
> I think there is a mistake in your line of thought 

Maybe, let's discuss tomorrow.

Reply:
It might be because of `GlobalSnapshotManager`, which performs "dispatching of dispatchings" (🫠). Try to change
```
internal actual val GlobalSnapshotManagerDispatcher: CoroutineDispatcher = MainUIDispatcher
```
to
```
internal actual val GlobalSnapshotManagerDispatcher: CoroutineDispatcher =
    Executors.newSingleThreadExecutor().asCoroutineDispatcher()
```

A concrete global dispatcher is implementation details. Swing dispatcher is only for avoiding bugs which aren't fixed yet.

Reply:
> A concrete global dispatcher is implementation details

Also, `GlobalSnapshotManager` only performs additional `sendApplyNotifications`. `sendApplyNotifications` can be also called from any thread by anyone. `derivedStateOf,`, for example called it. Try to add:
```
repeat(10000) {
  val c = derievedStateOf { color }
}
```
After
```
color = 
```

Reply:
Yes, it indeed causes this:

Source thread: Thread-5
Update thread: pool-1-thread-1 @coroutine#2
Source thread: Thread-5
Update thread: pool-1-thread-1 @coroutine#2
Source thread: Thread-5
Update thread: pool-1-thread-1 @coroutine#2
Source thread: Thread-5
Update thread: pool-1-thread-1 @coroutine#2
Source thread: Thread-5
Update thread: pool-1-thread-1 @coroutine#2
Source thread: Thread-5
Update thread: pool-1-thread-1 @coroutine#2

Reply:
Or just call `Snapshot.sendApplyNotifications ` after `color = `, it is valid to call it manually

Reply:
> Also, GlobalSnapshotManager only performs additional sendApplyNotifications. sendApplyNotifications can be also called from any thread by anyone. derivedStateOf, for example called it. Try to add:

What should I see after adding it? There is nothing suspicious.

Reply:
> Swing dispatcher is only for avoiding bugs which aren't fixed yet.

It sounds like a joke, but I think at the moment it looks like a nice feature.

Reply:
In this case we could simply make ALL calls to Swing interop to be executed in `SwingUtils.invokeLater`. I assume they have a fair ordering?

Reply:
> What should I see after adding it? There is nothing suspicious.

it is about seeing calls it another thread, multiple times. You already saw it [here](https://github.com/JetBrains/compose-multiplatform-core/pull/1489#discussion_r1706925740)

> It sounds like a joke, but I think at the moment it looks like a nice feature.

I was talking about the Swing dispatcher inside `GlobalSnapshotManager`

> where all changes to interop elements are delayed to be executed in a single transaction.

This was solved by `InteropViewUpdater` in the previous code, not sure what is used on iOS. Also, my suggestion about "custom dispatcher" is exactly about this - delay updates, and perform them in the end of `render`

Reply:
Okay, we need a separate skiko PR then

Reply:
https://github.com/JetBrains/skiko/pull/974

Reply:
https://github.com/JetBrains/compose-multiplatform-core/pull/1489/commits/a06a6a167241e4b98f513f67a02cd9c7208433e5 looks good to me 👍, except one additional comment.